Start reducing heap usage

kevans91 opened the pull request:
This takes some basic first steps to reducing task usage, and adds a task that prints out the high watermark of each of the tasks we're creating every 10 seconds. Ideally, we would enable this and run it through intense usage, then tune down the allocated stack for each task based on real-world data. The output values are minimum words left in the stack at the highest usage, so we should be able to roughly decrease the task stack directly by that amount minus a decent amount of wiggle room since I'm sure this will often differ from run to run.

I made the sampling task contingent on ``INCLUDE_uxTaskGetStackHighWaterMark` for easy enable/disable rather than having to remove it after the initial experiment is done, in case future tuning is desired.

kevans91 committed:
main: easy steps towards reducing heap usage

There's only going to ever be one ethernetTask/uart0Task run, so push these
large buffers off of the stack into .bss.  This moves 5KB into SRAM, but we
have enough slop to cope with it.

Start by reducing the amount of stack allocated for these tasks, 8192 words
is what ethernetTask had prior to these large buffers so let's just restore
that under the assumption it was bumped a lot and not necessarily
proportional to what was needed (4k / word size).

uart is reduced by 1024 *words* (not bytes), but I suspect even that is a
bit higher than needed.  We'll add a task to do some periodic stack sampling
shortly.

kevans91 committed:
main: add a task for stack sampling

Sample all tasks' stack high watermarks every, say, 10s, and write it out.

This should help us diagnose how much wiggle room we currently have, and
allow us to identify if we can reduce any of our estimates to save on heap
usage.

Ownasaurus commented on main.c:
```diff
@@ -383,6 +395,38 @@ void heartbeatTask(void *pvParameters)
+            if (stackWatermark == 0) {
+                task_print("STACK: %s: OVERFLOW\r\n", td->name);
+            } else {
+                task_print("STACK %s: %u words (%u bytes)\r\n", td->name, stackWatermark, stackWatermark / sizeof(StackType_t));
```
Would you kindly change that to `* sizeof` instead of `/ sizeof?` Thank you very much.

Ownasaurus commented:
Thank you very much for your time. This is fantastic and something I should have implemented months ago.

Ownasaurus commented on main.c:
```diff
@@ -80,14 +88,18 @@ int main(void)
+    creationResult = xTaskCreate(ethernetTask, (const portCHAR *)"ENET", 8192, NULL, 3, &ethernetHdl);
     ASSERT(creationResult == pdPASS);
-    creationResult = xTaskCreate(ethernetTask, (const portCHAR *)"ENET", 19456, NULL, 3, NULL);
+    creationResult = xTaskCreate(EXISendTask, (const portCHAR *)"EXISend", 8192, NULL, 4, &EXIHdl);
```
Runtime analysis shows that 2048 should be sufficient for this task.

Ownasaurus commented on main.c:
```diff
@@ -80,14 +88,18 @@ int main(void)
+    creationResult = xTaskCreate(uart0Task, (const portCHAR *)"UART0", 7168, NULL, 2, &uartHdl);
     ASSERT(creationResult == pdPASS);
-    creationResult = xTaskCreate(uart0Task, (const portCHAR *)"UART0", 8192, NULL, 2, NULL);
+    creationResult = xTaskCreate(ethernetTask, (const portCHAR *)"ENET", 8192, NULL, 3, &ethernetHdl);
```
Runtime analysis shows that 2048 should be sufficient for this task.

Ownasaurus commented on main.c:
```diff
@@ -80,14 +88,18 @@ int main(void)
-    creationResult = xTaskCreate(heartbeatTask, (const portCHAR *)"HB", 1024, NULL, 1, NULL);
+    creationResult = xTaskCreate(heartbeatTask, (const portCHAR *)"HB", 1024, NULL, 1, &heartbeatHdl);
+    ASSERT(creationResult == pdPASS);
+    creationResult = xTaskCreate(uart0Task, (const portCHAR *)"UART0", 7168, NULL, 2, &uartHdl);
```
Runtime analysis shows that 1024 should be sufficient for this task.

kevans91 committed:
main: add heap usage to our stack sampling task

kevans91 committed:
main: tune down task sizes based on collected data

Ownasaurus confirmed that at peak usage, these depths should be perfectly
fine with some wiggle room.  This should shave off about ~120k from
the existing heap usage.

Go ahead and disable the task stack instrumentation for now, to be
re-enabled as needed for diagnostics.

Ownasaurus commented:
Amazing work, @kevans91.
